lib/evaluate/character.h

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_EVALUATE_CHARACTER_H_
+#define FORTRAN_EVALUATE_CHARACTER_H_
+
+#include "type.h"
+#include <string>
+
+// character(1) is ISO/IEC 646:1991 (ASCII)
+//
+// character(2) is a variant of EUC-JP: The code points are the same as EUC-JP,
+// but the encoding has a two bytes fix-size (EUC-JP encoding has a variable
+// length). The one byte EUC-JP character representations are simply
+// zero-extended to two byte representations. The three bytes character
+// representation of EUC-JP (JIS X 0212) are not supported in this internal
+// encoding.
+//
+// character(4) is ISO/IEC 10646 UCS-4 (~ UTF-32)
+
+namespace Fortran::evaluate {
+
+template<int KIND> class CharacterUtils {
+  using Character = Scalar<Type<TypeCategory::Character, KIND>>;
+  using CharT = typename Character::value_type;
+
+public:
+  static constexpr bool IsValidCharacterCode(std::uint64_t code) {
+    if constexpr (KIND == 1) {
+      return IsValidASCII(code);
+    } else if constexpr (KIND == 2) {
+      return IsValidInternalEUC_JP(code);
+    } else if constexpr (KIND == 4) {
+      return IsValidUCS4(code);
+    } else {
+      static_assert(true, "bad character kind");
+    }
+  }
+
+  // CHAR also implements ACHAR
+  static Character CHAR(std::uint64_t code) {
+    return Character{{static_cast<CharT>(code)}};
+  }
+
+  // ICHAR also implements IACHAR
+  static int ICHAR(const Character &c) {
+    CHECK(c.length() == 1);
+    return c[0];
+  }
+
+  static Character NEW_LINE() { return Character{{NewLine()}}; }
+
+  static Character ADJUSTL(const Character &str) {
+    auto pos{str.find_first_not_of(Space())};
+    if (pos != Character::npos && pos != 0) {
+      return Character{str.substr(pos) + Character(pos, Space())};
+    }
+    // else empty or only spaces, or no leading spaces
+    return str;
+  }
+
+  static Character ADJUSTR(const Character &str) {
+    auto pos{str.find_last_not_of(Space())};
+    if (pos != Character::npos && pos != str.length() - 1) {
+      auto delta{str.length() - 1 - pos};
+      return Character{Character(delta, Space()) + str.substr(0, pos + 1)};
+    }
+    // else empty or only spaces, or no trailing spaces
+    return str;
+  }
+
+private:
+  static constexpr bool IsValidASCII(std::uint64_t code) { return code < 128; }
```

Comment:
This results in a warning if you use `ACHAR(129)`, for example. Why not accept any code < 256?

Reply:
I went for simple ASCII because the standard was not asking for more.
Also, currently the parser/semantic analysis of character literal with ASCII >= 128 creates a `character(kind=1, len=2)`. So `len("ÿ")` (ascii 255) returns 2 and so does `len("<80>") ` (ascii 128).
Which means that `achar(255).EQ.("ÿ")` would be `false` today if it was allowed by achar.

I think it is probably safe to support extended ASCII as 7-bit systems are probably not the target today, but then it requires modification in the parser or character literal analysis.
@klausler and @tskeith, should I modify this ?

Reply:
Other compilers that I tried all get `len("ÿ")` to be 2 and `len(achar(255))` to be 1. That's true for PGI, gfortran, ifort, nag, and xlf. None of them warn about `achar(255)`.

I find it surprising that `"ÿ"` is two characters but I guess that the few current Fortran programmers that do things like might not.

Reply:
0xFF is not a valid codepoints in UTF-8.

Do the 3 and 4 byte encodings of UTF-8 also report as LEN=2?

Interesting however that LEN reports the storage in bytes rather than the symbolic length of the string for all these implementation. My cursory look at the standard just led to "character length", which seems ambiguous in a superficial glance.

Reply:
`LEN` is length in characters.

`LEN(ACHAR())` is defined to be 1.

Quoted character literals of kind 1 can hold UTF-8 sequences or other non-ASCII source characters, but each byte is treated as a distinct 'character' in the literal.  This allows UTF-8 file names on OPEN statements to appear naturally as literals.  Character literals of kinds 2 and 4 undergo conversion from Kanji/UTF-8 to their 16/32-bit encodings.  The scanner actually cares about source encoding only when scanning Hollerith literals (the character count must be correctly interpreted).  We might have to support other source encodings in the future (e.g., Latin-1) and that will be easy so long as they're one-byte extensions of seven-bit ASCII.

Character variables (not quoted literals) of kind 1 can be abused to hold arbitrary byte sequences.  `ACHAR(I,kind=1)` should just take the least-order 8 bits of I; kinds 2 and 4 should take 16 and 32.  Or, in other words, `IACHAR(ACHAR())` and `ACHAR(IACHAR())` should be no-ops when the integer and character kind values are the same.  No UTF-8 semantics should apply to these.

Reply:
Thanks for your inputs. Checking more with other compilers behavior, and based on your comments it will bring more harm than good to do any checks related to the encoding here.

I have removed all validity checks in `char` and `achar` which now behave like pgfortran and xlf by simply keeping the lower bits fitting into the character underlying type.

I also changed a bit the signed/unsigned cast to ensure `ACHAR(IACHAR())` is a no-op when possible.

Reply:
You might consider an optional warning for values that are not in the range specified in 16.9.3 para 5.

Reply:
I agree it would be nice to do so but I am not comfortable regarding how to emit optional warnings in f18 so I would rather not emit a warning at all using the fact that 16.9.3 also says "otherwise, the result is processor dependent" when the value is outside of the specified range. f18 processor dependent behavior in this case is similar to pgfortran and xlf. I could also check a little bit that pgfortran kanji `ncharacter` `nchar(x)` works as f18 `char(x, 2)`, though pgfortran cannot fold `nchar` and I still do not know make `ncharacter` literals .

All that to say, if nobody has any other objections, I will merge this PR as it is now.